web_src/js/features/user-auth-webauthn.js

```diff
@@ -0,0 +1,199 @@
+      if (resp && resp['redirect']) {
+        window.location.href = resp['redirect'];
+      } else {
+        window.location.href = '/';
+      }
+    },
+    error: (xhr) => {
+      if (xhr.status === 500) {
+        webAuthnError('unknown');
+        return;
+      }
+      webAuthnError('unable-to-process');
+    }
+  });
+}
+
+// Encode an ArrayBuffer into a base64 string.
+function bufferEncode(value) {
+  return encode(value)
+    .replace(/\+/g, '-')
+    .replace(/\//g, '_')
+    .replace(/=/g, '');
+}
+
+function webauthnRegistered(newCredential) {
+  const attestationObject = new Uint8Array(newCredential.response.attestationObject);
+  const clientDataJSON = new Uint8Array(newCredential.response.clientDataJSON);
+  const rawId = new Uint8Array(newCredential.rawId);
+
+  return $.ajax({
+    url: `${appSubUrl}/user/settings/security/webauthn/register`,
+    type: 'POST',
+    headers: {'X-Csrf-Token': csrfToken},
+    data: JSON.stringify({
+      id: newCredential.id,
+      rawId: bufferEncode(rawId),
+      type: newCredential.type,
+      response: {
+        attestationObject: bufferEncode(attestationObject),
+        clientDataJSON: bufferEncode(clientDataJSON),
+      },
+    }),
+    dataType: 'json',
+    contentType: 'application/json; charset=utf-8',
+  }).then(() => {
+    window.location.reload();
+  }).fail((xhr) => {
+    if (xhr.status === 409) {
+      webAuthnError('duplicated');
+      return;
+    }
+    webAuthnError('unknown');
+  });
+}
+
+function webAuthnError(errorType, message) {
+  $('#webauthn-error [data-webauthn-error-msg]').hide();
+  if (errorType === 0 && message && message.length > 1) {
+    $(`#webauthn-error [data-webauthn-error-msg=0]`).text(message);
+    $(`#webauthn-error [data-webauthn-error-msg=0]`).show();
+  } else {
+    $(`#webauthn-error [data-webauthn-error-msg=${errorType}]`).show();
+  }
+  $('#webauthn-error').modal('show');
+}
+
+function detectWebAuthnSupport() {
+  if (typeof window.PublicKeyCredential !== 'function') {
+    $('#register-button').prop('disabled', true);
+    $('#login-button').prop('disabled', true);
+    webAuthnError('browser');
+    return false;
+  }
+
+  if (window.location.protocol === 'http:' && (window.location.hostname !== 'localhost' && window.location.hostname !== '127.0.0.1')) {
+    $('#register-button').prop('disabled', true);
+    $('#login-button').prop('disabled', true);
+    webAuthnError('insecure');
+    return false;
+  }
+  return true;
+}
+
+export function initUserAuthWebAuthnRegister() {
+  if ($('#register-webauthn').length === 0) {
+    return;
+  }
+
+  if (!detectWebAuthnSupport()) {
+    return;
+  }
+
+  $('#register-device').modal({allowMultiple: false});
+  $('#webauthn-error').modal({allowMultiple: false});
+  $('#register-webauthn').on('click', (e) => {
+    e.preventDefault();
+    webAuthnRegisterRequest();
+  });
+}
+
+function bufferDecode(value) {
+  return Uint8Array.from(atob(value), (c) => c.codePointAt(0));
+}
```

Comment:
his can be replaced with `decode` from `uint8-to-base64`:

https://github.com/WebReflection/uint8-to-base64/blob/89ebf6dc1090a4e4e33318b5d747e9006f3c7b2c/esm/index.js#L12